### DISCUSSION TITLE
Protocol without Protocol Buffer: gun -> guna

### DISCUSSION THREAD

**Discussion by DuckSoft** [Qv2ray/gun-lite#6] [Mind Blown]
### Basic Ideas
The parsing overhead for LEB128 is huge. So why not remove it? 

According to https://github.com/grpc/grpc-go/blob/master/Documentation/encoding.md#codecs-serialization-and-deserialization, we know that we can define our own Codecs. For the most simple use case, we can just copy the packet as-is, without having the need to calculate the length.

### Structure
The knowledge above enables us to turn the structure from (original `gun` protocol):

```
    gRPC Header              gRPC Payload (Protobuf)
┌────┬─────────────┐ ┌────┬────────────────┬────────────────┐
│ 00 │ ?? ?? ?? ?? │ │ 0A │ ?? .. .. .. .. │ ?? .. .. .. .. │
└────┴─────────────┘ └────┴────────────────┴────────────────┘
  ▲    ▲                             ▲       ▲
  │    └── gRPC Payload Length       │       │
  │        (Fixed 4 bytes)           │       └── Data
  │                                  │
  └─────── Compression Identifier    └────────── Data Length
           (No Compression = 0x00)               (ULEB128)
```

to the following:
```
    gRPC Header            Data
┌────┬─────────────┐ ┌────────────────┐
│ 00 │ ?? ?? ?? ?? │ │ ?? .. .. .. .. │
└────┴─────────────┘ └────────────────┘
  ▲    ▲
  │    └── Data Length in DWORD
  │
  └─────── No Compression Flag
```
which, I would like to name as **`guna` protocol**.

### Benefit
- More Lightweight
- No more LEB128 stuff needed
- No need to mess with Protocol Buffers

### Drawback
- Lose Protocol Buffers Benefit

### Acknowledgements
The idea is not mine. I am asking for the permission to put his name here.

**Comment by Loyalsoldier** [comment]
What if the data length is larger than `2^32`?

**Reply by DuckSoft** [reply]
You will probably not have a single packet with a size bigger than 4GiB...
